experimental/segmenter/src/complex.rs

```diff
@@ -230,42 +242,39 @@ pub(crate) fn complex_language_segment_utf16(
 
 /// Return UTF-8 segment offset array using dictionary or lstm segmenter.
 #[allow(unused_variables)]
-pub(crate) fn complex_language_segment_str(
-    dictionary: Option<&Dictionary>,
-    lstm: Option<&LstmPayloads>,
-    grapheme: Option<&RuleBreakDataV1>,
-    input: &str,
-) -> Vec<usize> {
+pub(crate) fn complex_language_segment_str(payloads: &ComplexPayloads, input: &str) -> Vec<usize> {
     let mut result: Vec<usize> = Vec::new();
     let lang_iter = LanguageIterator::new(input);
     let mut offset = 0;
-    for str_per_lang in lang_iter {
-        if let Some(first_ch) = str_per_lang.chars().next() {
+    for (str_per_lang, lang) in lang_iter {
+        if lang == Language::Unknown {
+            offset += str_per_lang.len();
+            result.push(offset);
+        } else if let Some(lstm) = payloads.select_lstm(lang) {
             #[cfg(feature = "lstm")]
-            if let Some(lstm) = lstm {
-                if let Some(model) = lstm.best(first_ch as u32) {
-                    if let Ok(segmenter) = LstmSegmenter::try_new(model, grapheme) {
-                        let breaks = segmenter.segment_str(str_per_lang);
-                        result.extend(breaks.map(|n| offset + n));
-                        offset += str_per_lang.len();
-                        result.push(offset);
-                        continue;
-                    }
-                }
-            }
-
-            if let Some(dictionary) = dictionary {
-                if let Some(grapheme) = grapheme {
-                    if let Some(payload) = dictionary.best(first_ch as u32) {
-                        if let Ok(segmenter) = DictionarySegmenter::try_new(payload, grapheme) {
-                            let breaks = segmenter.segment_str(str_per_lang);
-                            result.extend(breaks.map(|n| offset + n));
-                            offset += str_per_lang.len();
-                            continue;
-                        }
-                    }
-                }
+            {
+                let segmenter = crate::lstm::LstmSegmenter::new(lstm, &payloads.grapheme);
+                let breaks = segmenter.segment_str(str_per_lang);
+                result.extend(breaks.map(|n| offset + n));
+                offset += str_per_lang.len();
             }
+        } else if let Some(dict) = payloads.select_dict(lang) {
+            let segmenter = DictionarySegmenter::new(dict, &payloads.grapheme);
+            let breaks = segmenter.segment_str(str_per_lang);
+            result.extend(breaks.map(|n| offset + n));
+            offset += str_per_lang.len();
+        } else {
+            // Create error for logging
+            DataError::custom("No segmentation model for language").with_display_context(
```

Comment:
Thought: Actually I kind-of like the idea of performing logging through the icu_provider crate. It keeps all the needed machinery in one place. We should probably add our own logging macro that we export from icu_provider.